File: src/ch03/code_5.py
from typing import List

from rich import print


def exhaustive_search(nums: List[int], target: int) -> bool:
    for bit in range(1<<len(nums)):
        total: int = 0
        for i in range(len(nums)):
            if bit & (1<<i):
                total += nums[i]
        if total == target:
            return True
    
    return False


if __name__ == '__main__':
    nums: List[int] = [1, 2, 4, 5, 11]
    target: int = 10
    result: bool = exhaustive_search(nums=nums, target=target)
    
    print(result)

File: src/ch03/code_5.rs
fn exhaustive_search(nums: Vec<i32>, target: i32) -> bool {
    for bit in 0..(1<<nums.len()) {
        let mut total: i32 = 0;
        for (idx, num) in nums.iter().enumerate() {
            if bit & (1<<idx) > 0 {
                total += *num;
            }
        }
        if total == target {
            return true;
        }
    }
    false
}

fn main() {
    let nums: Vec<i32> = vec![1, 2, 4, 5, 11];
    let target: i32 = 10;
    let result: bool = exhaustive_search(nums, target);

    println!("{}", result);
}